flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java

```diff
@@ -126,63 +196,116 @@ void pollMessageAfterTimeout(PulsarPartitionSplitReaderBase<String> splitReader)
-            this.splitReader = checkNotNull(splitReader);
-        }
+    @Test
+    void emptyTopicWithoutSeek() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        String topicName = randomAlphabetic(10);
+        operator().createTopic(topicName, DEFAULT_PARTITIONS);
+        handleSplit(splitReader, topicName, 0);
+        fetchedMessages(splitReader, 0, true);
+    }
 
-        @Override
-        public String getDisplayName(int invocationIndex) {
-            return splitReader.getClass().getSimpleName();
+    @Test
+    void wakeupSplitReaderShouldNotCauseException() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        handleSplit(splitReader, "non-exist", 0);
+        AtomicReference<Throwable> error = new AtomicReference<>();
+        Thread t =
+                new Thread(
+                        () -> {
+                            try {
+                                splitReader.fetch();
+                            } catch (Throwable e) {
+                                error.set(e);
+                            }
+                        },
+                        "testWakeUp-thread");
+        t.start();
+        long deadline = System.currentTimeMillis() + 5000L;
+        while (t.isAlive() && System.currentTimeMillis() < deadline) {
+            splitReader.wakeUp();
+            sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
         }
+        assertNull(error.get());
+    }
 
-        @Override
-        public List<Extension> getAdditionalExtensions() {
-            return Collections.singletonList(
-                    new ParameterResolver() {
-                        @Override
-                        public boolean supportsParameter(
-                                ParameterContext parameterContext,
-                                ExtensionContext extensionContext)
-                                throws ParameterResolutionException {
-                            return parameterContext
-                                    .getParameter()
-                                    .getType()
-                                    .equals(PulsarPartitionSplitReaderBase.class);
-                        }
-
-                        @Override
-                        public Object resolveParameter(
-                                ParameterContext parameterContext,
-                                ExtensionContext extensionContext)
-                                throws ParameterResolutionException {
-                            return splitReader;
-                        }
-                    });
-        }
+    @Test
+    void assignNoSplits() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        assertNull(fetchedMessage(splitReader));
     }
+
+    /** Create a split reader with max message 1, fetch timeout 1s. */
+    protected abstract PulsarPartitionSplitReaderBase<String> splitReader();
```

Comment:
You can implement an extension to instantiate the test SplitReader in the subclasses and directly passes it to the test without the need of an abstract method. You can take a look at how the `SourceTestSuiteBase` does it with the `TestEnvironment` [1]

[1] https://github.com/apache/flink/blob/master/flink-test-utils-parent/flink-connector-testing/src/main/java/org/apache/flink/connectors/test/common/testsuites/SourceTestSuiteBase.java

Reply:
Do you mean use @TestTemplate and a custom TestInvocationContextProvider ? Me and @syhily used @TestTemplate and TestInvocationContext and then refactored to what is looks like now (cause we thought they were an overkill in our case).  But sure I can switch back to use the TestInvocationContextProvider for extensibility.